@xen-orchestra/backups-cli/index.js

```diff
@@ -0,0 +1,354 @@
+const { pipe, promisifyAll } = require('promise-toolbox')
+
+const fs = promisifyAll(require('fs'))
+const handler = require('@xen-orchestra/fs').getHandler({ url: 'file://' })
+
+// -----------------------------------------------------------------------------
+
+const asyncMap = curryRight((iterable, fn) =>
+  Promise.all(
+    Array.isArray(iterable) ? iterable.map(fn) : Array.from(iterable, fn)
+  )
+)
+
+const filter = (...args) => thisArg => thisArg.filter(...args)
+
+// TODO: better check?
+
+// our heuristic is not good enough, there has been some false positives
+// (detected as invalid by us but valid by `tar` and imported with success),
+// either:
+// - these files were normal but the check is incorrect
+// - these files were invalid but without data loss
+// - these files were invalid but with silent data loss
+//
+// maybe reading the end of the file looking for a file named
+// /^Ref:\d+/\d+\.checksum$/ and then validating the tar structure from it
+//
+// https://github.com/npm/node-tar/issues/234#issuecomment-538190295
+const isValidTar = async path => {
+  const fd = await fs.open(path, 'r')
+  try {
+    const { size } = await fs.fstat(fd)
+    if (size <= 1024 || size % 512 !== 0) {
+      return false
+    }
+
+    const buf = Buffer.allocUnsafe(1024)
+    assert.strictEqual(
+      await fs.read(fd, buf, 0, buf.length, size - 1 - buf.length),
+      buf.length
+    )
+    return buf.every(_ => _ === 0)
+  } finally {
+    fs.close(fd).catch(noop)
+  }
+}
+
+const noop = Function.prototype
+
+const readDir = path =>
+  fs.readdir(path).then(
+    entries => {
+      entries.forEach((entry, i) => {
+        entries[i] = `${path}/${entry}`
+      })
+
+      return entries
+    },
+    () => []
+  )
+
+// -----------------------------------------------------------------------------
+
+// chain is an array of VHD from child to parent
+//
+// the whole chain will be merged into parent, parent will be renamed to child
+// and all the others will deleted
+async function mergeVhdChain(chain) {
+  assert(chain.length >= 2)
+
+  const child = chain[0]
+  const parent = chain[chain.length - 1]
+  const children = chain.slice(0, -1).reverse()
+
+  console.warn('Unused parents of VHD', child)
+  chain
+    .slice(1)
+    .reverse()
+    .forEach(parent => {
+      console.warn(' ', parent)
+    })
+  force && console.warn('  merging…')
+  console.warn('')
+  if (force) {
+    // `mergeVhd` does not work with a stream, either
+    // - makes it accept a stream
+    // - or create synthetic VHD which is not a stream
+    throw new Error('FIXME')
+    // await mergeVhd(
+    //   handler,
+    //   parent,
+    //   handler,
+    //   children.length === 1
+    //     ? child
+    //     : await createSyntheticStream(handler, children)
+    // )
+  }
+
+  await Promise.all([
+    fs.rename(parent, child),
+    asyncMap(children.slice(0, -1), child => {
+      console.warn('Unused VHD', child)
```

Comment:
Yes and no, it ties together the file and the action.
Otherwise it's difficult to say which files are deleted, especially with asynchronicity.

I'm open to better solution.